roles/neutron-common/templates/usr/local/bin/neutron-reset

```diff
@@ -61,33 +62,69 @@ class Interface(object):
             return True
 
         # veth interfaces named tap... belong to Neutron
-        if re.match(r'^tap[0-9a-z]{8}-[0-9a-z]{2}$', self.ifname) and \
-                self.iftype() == 'veth':
-            return True
+        if re.match(r'^tap[0-9a-z]{8}-[0-9a-z]{2}$', self.ifname) and self.iftype() == 'veth':
+               return True
 
-        if re.match(r'^vxlan[\d]{1-4}$', self.ifname) and \
-                self.iftype() == 'vxlan':
+        if re.match(r'^vxlan\-[0-9]{1,4}$', self.ifname) and self.iftype() == 'vxlan':
             return True
 
-        if re.match(r'^brq[0-9a-z]{8}-[0-9a-z]{2}$', self.ifname) and \
-                self.iftype() == 'bridge':
+        if re.match(r'^brq[0-9a-z]{8}-[0-9a-z]{2}$', self.ifname) and self.iftype() == 'bridge':
             return True
 
         return False
 
+    @classmethod
+    def get_nova_interface(self, namespace=None):
+        cmd = ['ip', '-d', 'link', 'show']
+        if namespace:
+            cmd = ['ip', 'netns', 'exec', namespace] + cmd
+
+        output = subprocess.check_output(cmd)
+
+        pattern = re.compile(r'^\d+: ([^:]+): ')
+        iface_list = list()
+
+        for line in output.splitlines():
+            match = pattern.match(line)
+            if match:
+                iface = Interface(match.group(1), namespace)
+
+                if iface.is_nova_interface():
+                    iface_list.append(iface)
+
+        return iface_list
+
+    def is_nova_interface(self):
+        return self.ifname.startswith('tap') and self.iftype() == 'tun'
+
     def iftype(self):
         if self._iftype is None:
             self._fetch_additional_info()
 
         return self._iftype
 
+    def attached_bridge(self):
+        if self._bridge_attached is None:
+            self._fetch_additional_info()
+
+        return self._bridge_attached
+
+    def wait_for_link_up(self):
+        cmd = self._ip_cmd() + ['link', 'show', self.ifname]
+
+        output = os.popen(' '.join([x for x in cmd])).read()
+
+        while "state UP" not in output:
+            output = os.popen(' '.join([x for x in cmd])).read()
```

Comment:
subprocess.checked_output would be better than popen().read() -- since the later may not return the complete output.

popen() is good when you need to stream the output of a long running command, but subprocess module does a very nice job wrapping the common use case of commands which quickly return.